File: merge_intervals/Merge_Intervals.cpp
class Merge_Intervals {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        if (intervals.empty() || intervals[0].empty()) {
            return vector<vector<int>>(1, vector<int>());
        }
        
        map<int, int> segments;
        for (auto& interval : intervals) {
            int start = interval[0];
            int end = interval[1];
            map<int, int>::iterator floor = segments.lower_bound(start);
            if (floor != segments.begin()) {
                floor--;
                if (floor->second >= start) {
                    end = max(end, floor->second);
                    start = floor->first;
                    segments.erase(floor->first);
                }
            }
            
            map<int, int>::iterator ceil = segments.lower_bound(start);
            while (ceil != segments.end() && ceil->first <= end) {
                end = max(end, ceil->second);
                segments.erase(ceil->first);
                ceil = segments.upper_bound(start);
            }
            
            segments.insert({start, end});
        }
        
        vector<vector<int>> merged_intervals(segments.size(), vector<int>(2));
        int idx = 0;
        for (auto& interval : segments) {
            merged_intervals[idx++] = {interval.first, interval.second};
        }
        
        return merged_intervals;
    }
};

File: merge_intervals/MergeIntervals.java
class MergeIntervals {
    public int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length == 0 || intervals[0] == null || intervals[0].length == 0) {
            return new int[1][0];
        }
        
        TreeMap<Integer, Integer> segments = new TreeMap<>();
        for (int[] interval : intervals) {
            int start = interval[0];
            int end = interval[1];
            Map.Entry<Integer, Integer> floor = segments.floorEntry(start);
            if (floor != null && floor.getValue() >= start) {
                segments.remove(floor.getKey());
                start = floor.getKey();
                end = Math.max(end, floor.getValue());
            }
            
            Map.Entry<Integer, Integer> ceil = segments.ceilingEntry(start);
            while (ceil != null && ceil.getKey() <= end) {
                segments.remove(ceil.getKey());
                end = Math.max(end, ceil.getValue());
                ceil = segments.ceilingEntry(start);
            }
            
            segments.put(start, end);
        }
        
        int[][] mergedIntervals = new int[segments.size()][];
        int idx = 0;
        for (Map.Entry<Integer, Integer> interval : segments.entrySet()) {
            mergedIntervals[idx++] = new int[]{interval.getKey(), interval.getValue()};
        }
        
        return mergedIntervals;
    }
}